lib/utils/diagnostics/latency/monitor.go

```diff
@@ -0,0 +1,339 @@
+	}
+
+	if c.PingInterval <= 0 {
+		c.PingInterval = 8 * time.Second
+	}
+
+	if c.ReportInterval <= 0 {
+		c.ReportInterval = 10 * time.Second
+	}
+
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+
+	return nil
+}
+
+const (
+	pingKey      = "ping-ticker"
+	reportingKey = "reporting-ticker"
+)
+
+var (
+	seventhJitter = retryutils.NewShardedSeventhJitter()
+	fullJitter    = retryutils.NewShardedFullJitter()
+	halfJitter    = retryutils.NewShardedHalfJitter()
+)
+
+// NewMonitor creates an unstarted [Monitor] with the provided configuration. To
+// begin sampling connection latencies [Monitor.Run] must be called.
+func NewMonitor(cfg MonitorConfig) (*Monitor, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ticker := interval.NewMulti(
+		cfg.Clock,
+		interval.SubInterval[string]{
+			Key:           pingKey,
+			FirstDuration: fullJitter(500 * time.Millisecond),
+			Jitter:        seventhJitter,
+			Duration:      cfg.PingInterval,
+		},
+		interval.SubInterval[string]{
+			Key:           reportingKey,
+			FirstDuration: halfJitter(1500 * time.Millisecond),
+			Jitter:        seventhJitter,
+			Duration:      cfg.ReportInterval,
+		},
+	)
+
+	return &Monitor{
+		clientPinger: cfg.ClientPinger,
+		serverPinger: cfg.ServerPinger,
+		reporter:     cfg.Reporter,
+		ticker:       ticker,
+		clock:        cfg.Clock,
+	}, nil
+}
+
+// GetStats returns a copy of the last known latency measurements.
+func (m *Monitor) GetStats() Statistics {
+	return Statistics{
+		Client: m.clientLatency.Load(),
+		Server: m.serverLatency.Load(),
+	}
+}
+
+// Run periodically records round trip times. It should not be called
+// more than once.
+func (m *Monitor) Run(ctx context.Context) {
+	defer m.ticker.Stop()
+
+	clientC, serverC := make(chan time.Time, 1), make(chan time.Time, 1)
+	go m.pingLoop(ctx, clientC, m.clientPinger, &m.clientLatency)
+	go m.pingLoop(ctx, serverC, m.serverPinger, &m.serverLatency)
+
+	for {
+		select {
+		case tick := <-m.ticker.Next():
+			switch tick.Key {
+			case pingKey:
+				// Ping the client
+				select {
+				case clientC <- tick.Time:
+				case <-ctx.Done():
+					return
+				default:
+				}
+
+				// Ping the server
+				select {
+				case serverC <- tick.Time:
+				case <-ctx.Done():
+					return
+				default:
+				}
+			case reportingKey:
+				if err := m.reporter.Report(ctx, m.GetStats()); err != nil {
+					log.WithError(err).Warn("failed to report latency stats")
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (m *Monitor) pingLoop(ctx context.Context, pingC <-chan time.Time, pinger Pinger, latency *atomic.Int64) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case then := <-pingC:
+			if err := pinger.Ping(ctx); err != nil {
+				log.WithError(err).Warn("unexpected failure sending ping")
+			} else {
+				latency.Store(m.clock.Now().Sub(then).Milliseconds())
+			}
```

Comment:
nit: using the tick time rather than a start time collected immediately before `Ping()` is called might throw off the data a bit.  The tick time was generated by the ticker goroutine, sent across to the `Run()` goroutine, which then sent it here.  If teleport is running cpu-constrained all that context-switching might be inflating the latency a bit.